content/lectures/lec00.md

```diff
@@ -6,4 +6,144 @@ assignments: []
+| Used for read-only operations             | Used to modify a resource | Used to create or replace a resource |                     |
+| Uses Query parameters to hold information |                           | Is idempotent                        |                     |
+| Is idempotent                             |                           |                                      |                     |
+| Ex. Retrieving a webpage                  | Ex. Submitting a form     | Ex. Uploading a file                 | Ex. Deleting a file |
+{{</table>}}
+
+2. URL - specifices what resource to request
+
+    For example: https://httpstat.us/200
+
+    https:// - **Protocol** normally http or https (unencrypted or use TLS encryption).
+
+    httpstat.us - **Hostname** the unique identifier of where to connect to.
+
+    200 - **Path** What resource on the host to interact with.
+
+3. Headers
+
+    HTTP headers allow the client to attach additional information to their request. Some common headers are:
+    * Authentication (`Authorization`)
+    * Caching (`Ages`, `Expires`)
+    * Cookies
+    * Body information (`Content-Type`, `Content-Length`)
+
+4. Body - data associated with the request
+
+    Of the HTTP methods we've covered, a body is only used in POST and PUT.
+    A body just consists of additional data that's sent as part of the request. This data is often a single file, form data, or a JSON object that the client sends to the server.
+
+### HTTP Response components
+
+TODO:
+
+## Demos
+
+### Browser
+
+The most common way to interact with an HTTP server is through the browser -- this is something we all do everyday. Let's open up a new browser window and see if we can get any more information about the actual HTTP requests underlying the page we view.
+
+Navigate to our course site: [cis188.github.io](https://cis188.github.io/), when we load the page lots of HTTP requests are fired off to retrieve everything that your browser is going to need to compile and load this page on your computer. We can actually examine these HTTP requests by right clicking on the window and selecting `inspect element` (on certain browsers, you might have to search through the help bar or configure your browser settings to allow developer tools). In your developer tool, you will be able to see a series of requests (normally under the `network` tab).
+
+![browser_screenshot](/demos/browser_screenshot.png)
+
+These are all HTTP requests that sent out by your browser. You can see the the duration of each request, the status code in the response, as well as the size and type of the response data. If you select a particular request, you can see the request and response headers (metadata passed alongside the actual data being transferred). All of this information can be helpful with debugging a website that is not loading or working as expected.
+
+### cURL
```

Comment:
An example of how to pass headers with curl would be a good demo. Also, this website is helpful for getting the status codes you want:
http://httpstat.us/

Reply:
@cphalen

Reply:
Agreed it would be good to actually specify a header. I'll add this to curl demo and might be useful for the later demos, too.